File: libs/shared-models/src/platform-data.ts
import { NavigationalStatus } from './common';

/** AIS maritime identification and voyage data (per UC2 AISDataType) */
export interface AISData {
  // Identity
  mmsi: string;
  imo?: string;
  callsign?: string;
  vesselName?: string;

  // Classification
  shipType?: number;
  shipTypeName?: string;
  flag?: string;

  // Voyage
  destination?: string;
  eta?: string;
  draught?: number;

  // Dimensions (UC2: dimensionA-D from reference point)
  dimensionA?: number;
  dimensionB?: number;
  dimensionC?: number;
  dimensionD?: number;
  lengthOverall?: number;
  beam?: number;

  // Navigation
  navStatus?: NavigationalStatus;
  rateOfTurn?: number;
  speedOverGround?: number;
  courseOverGround?: number;
  trueHeading?: number;
  positionAccuracyHigh?: boolean;
  specialManoeuvre?: boolean;

  // Message context
  messageType?: number;
  repeatIndicator?: number;
}

/** ADS-B / IFF transponder data (per UC2 IFFDataType + ModeSType + Mode5Type) */
export interface ADSBData {
  // Core identity
  icaoHex: string;
  registration?: string;
  aircraftType?: string;
  aircraftTypeName?: string;
  operatorIcao?: string;
  operatorName?: string;

  // Transponder codes (UC2 IFFDataType)
  squawk?: string;
  emergency?: string;
  mode1?: string;
  mode2?: string;

  // Mode S data (UC2 ModeSType)
  aircraftId?: string;
  flightAirborne?: boolean;
  indicatedAirSpeed?: number;
  trueAirSpeed?: number;
  groundSpeed?: number;
  magneticHeading?: number;

  // Mode 5 data (UC2 Mode5Type — military IFF)
  mode5FigureOfMerit?: number;
  nationalOriginCode?: number;
  missionCode?: number;

  // Position/altitude
  altitudeBaro?: number;
  altitudeGeom?: number;
  verticalRate?: number;
  onGround?: boolean;

  // Quality indicators
  category?: string;
  nacP?: number;
  nacV?: number;
  sil?: number;
  silType?: string;
  nic?: number;
  rc?: number;
  gva?: number;
  sda?: number;
}

/** Satellite TLE orbital data */
export interface TLEData {
  // Identity
  noradId: number;
  intlDesignator?: string;
  satName?: string;

  // TLE elements
  line1: string;
  line2: string;
  epoch?: string;

  // Orbital elements (derived)
  inclination?: number;
  eccentricity?: number;
  raan?: number;
  argOfPerigee?: number;
  meanAnomaly?: number;
  meanMotion?: number;
  period?: number;
  apogee?: number;
  perigee?: number;

  // Classification
  objectType?: string;
  rcsSize?: string;
  launchDate?: string;
  decayDate?: string;
  country?: string;
}

/** Link 16 / JREAP-C tactical data link */
export interface Link16Data {
  trackNumber: number;
  jSeriesLabel: string;
  originatingUnit?: string;
  quality?: number;
  exerciseIndicator?: boolean;
  simulationIndicator?: boolean;
  forceIdentity?: string;
}

/** Cursor on Target metadata */
export interface CoTData {
  uid: string;
  cotType: string;
  how?: string;
  ce?: number;
  le?: number;
  staleTime?: string;
  accessControl?: string;
  opex?: string;
  qos?: string;
}

/** UAV-specific telemetry data (per UC2 UAVDataType) */
export interface UAVData {
  make?: string;
  model?: string;
  serialNumber?: string;
  macAddress?: string;
  operatingFrequency?: number;
  frequencyRange?: number;
}

/** Discriminated union — only one populated per entity */
export interface PlatformData {
  ais?: AISData;
  adsb?: ADSBData;
  tle?: TLEData;
  link16?: Link16Data;
  cot?: CoTData;
  uav?: UAVData;
}

File: libs/common-go/models/platform_data.go
package models

// PlatformData is a discriminated union — only one field populated per entity.
type PlatformData struct {
	AIS    *AISData    `json:"ais,omitempty"`
	ADSB   *ADSBData   `json:"adsb,omitempty"`
	TLE    *TLEData    `json:"tle,omitempty"`
	Link16 *Link16Data `json:"link16,omitempty"`
	CoT    *CoTData    `json:"cot,omitempty"`
	UAV    *UAVData    `json:"uav,omitempty"`
}

// AISData holds AIS maritime identification and voyage data.
type AISData struct {
	MMSI                 string  `json:"mmsi"`
	IMO                  *string `json:"imo,omitempty"`
	Callsign             *string `json:"callsign,omitempty"`
	VesselName           *string `json:"vesselName,omitempty"`
	ShipType             *int    `json:"shipType,omitempty"`
	ShipTypeName         *string `json:"shipTypeName,omitempty"`
	Flag                 *string `json:"flag,omitempty"`
	Destination          *string `json:"destination,omitempty"`
	ETA                  *string `json:"eta,omitempty"`
	Draught              *float64 `json:"draught,omitempty"`
	DimensionA           *float64 `json:"dimensionA,omitempty"`
	DimensionB           *float64 `json:"dimensionB,omitempty"`
	DimensionC           *float64 `json:"dimensionC,omitempty"`
	DimensionD           *float64 `json:"dimensionD,omitempty"`
	LengthOverall        *float64 `json:"lengthOverall,omitempty"`
	Beam                 *float64 `json:"beam,omitempty"`
	NavStatus            *string  `json:"navStatus,omitempty"`
	RateOfTurn           *float64 `json:"rateOfTurn,omitempty"`
	SpeedOverGround      *float64 `json:"speedOverGround,omitempty"`
	CourseOverGround     *float64 `json:"courseOverGround,omitempty"`
	TrueHeading          *float64 `json:"trueHeading,omitempty"`
	PositionAccuracyHigh *bool    `json:"positionAccuracyHigh,omitempty"`
	SpecialManoeuvre     *bool    `json:"specialManoeuvre,omitempty"`
	MessageType          *int     `json:"messageType,omitempty"`
	RepeatIndicator      *int     `json:"repeatIndicator,omitempty"`
}

// ADSBData holds ADS-B / IFF transponder data.
type ADSBData struct {
	ICAOHex           string   `json:"icaoHex"`
	Registration      *string  `json:"registration,omitempty"`
	AircraftType      *string  `json:"aircraftType,omitempty"`
	AircraftTypeName  *string  `json:"aircraftTypeName,omitempty"`
	OperatorICAO      *string  `json:"operatorIcao,omitempty"`
	OperatorName      *string  `json:"operatorName,omitempty"`
	Squawk            *string  `json:"squawk,omitempty"`
	Emergency         *string  `json:"emergency,omitempty"`
	Mode1             *string  `json:"mode1,omitempty"`
	Mode2             *string  `json:"mode2,omitempty"`
	AircraftID        *string  `json:"aircraftId,omitempty"`
	FlightAirborne    *bool    `json:"flightAirborne,omitempty"`
	IndicatedAirSpeed *float64 `json:"indicatedAirSpeed,omitempty"`
	TrueAirSpeed      *float64 `json:"trueAirSpeed,omitempty"`
	GroundSpeed       *float64 `json:"groundSpeed,omitempty"`
	MagneticHeading   *float64 `json:"magneticHeading,omitempty"`
	Mode5FigureOfMerit *int    `json:"mode5FigureOfMerit,omitempty"`
	NationalOriginCode *int    `json:"nationalOriginCode,omitempty"`
	MissionCode       *int     `json:"missionCode,omitempty"`
	AltitudeBaro      *float64 `json:"altitudeBaro,omitempty"`
	AltitudeGeom      *float64 `json:"altitudeGeom,omitempty"`
	VerticalRate      *float64 `json:"verticalRate,omitempty"`
	OnGround          *bool    `json:"onGround,omitempty"`
	Category          *string  `json:"category,omitempty"`
	NacP              *int     `json:"nacP,omitempty"`
	NacV              *int     `json:"nacV,omitempty"`
	SIL               *int     `json:"sil,omitempty"`
	SILType           *string  `json:"silType,omitempty"`
	NIC               *int     `json:"nic,omitempty"`
	RC                *int     `json:"rc,omitempty"`
	GVA               *int     `json:"gva,omitempty"`
	SDA               *int     `json:"sda,omitempty"`
}

// TLEData holds satellite TLE orbital data.
type TLEData struct {
	NoradID         int      `json:"noradId"`
	IntlDesignator  *string  `json:"intlDesignator,omitempty"`
	SatName         *string  `json:"satName,omitempty"`
	Line1           string   `json:"line1"`
	Line2           string   `json:"line2"`
	Epoch           *string  `json:"epoch,omitempty"`
	Inclination     *float64 `json:"inclination,omitempty"`
	Eccentricity    *float64 `json:"eccentricity,omitempty"`
	RAAN            *float64 `json:"raan,omitempty"`
	ArgOfPerigee    *float64 `json:"argOfPerigee,omitempty"`
	MeanAnomaly     *float64 `json:"meanAnomaly,omitempty"`
	MeanMotion      *float64 `json:"meanMotion,omitempty"`
	Period          *float64 `json:"period,omitempty"`
	Apogee          *float64 `json:"apogee,omitempty"`
	Perigee         *float64 `json:"perigee,omitempty"`
	ObjectType      *string  `json:"objectType,omitempty"`
	RCSSize         *string  `json:"rcsSize,omitempty"`
	LaunchDate      *string  `json:"launchDate,omitempty"`
	DecayDate       *string  `json:"decayDate,omitempty"`
	Country         *string  `json:"country,omitempty"`
}

// Link16Data holds Link 16 / JREAP-C tactical data link info.
type Link16Data struct {
	TrackNumber         int     `json:"trackNumber"`
	JSeriesLabel        string  `json:"jSeriesLabel"`
	OriginatingUnit     *string `json:"originatingUnit,omitempty"`
	Quality             *int    `json:"quality,omitempty"`
	ExerciseIndicator   *bool   `json:"exerciseIndicator,omitempty"`
	SimulationIndicator *bool   `json:"simulationIndicator,omitempty"`
	ForceIdentity       *string `json:"forceIdentity,omitempty"`
}

// CoTData holds Cursor on Target metadata.
type CoTData struct {
	UID           string  `json:"uid"`
	CoTType       string  `json:"cotType"`
	How           *string `json:"how,omitempty"`
	CE            *float64 `json:"ce,omitempty"`
	LE            *float64 `json:"le,omitempty"`
	StaleTime     *string  `json:"staleTime,omitempty"`
	AccessControl *string  `json:"accessControl,omitempty"`
	Opex          *string  `json:"opex,omitempty"`
	QOS           *string  `json:"qos,omitempty"`
}

// UAVData holds UAV-specific telemetry data.
type UAVData struct {
	Make               *string  `json:"make,omitempty"`
	Model              *string  `json:"model,omitempty"`
	SerialNumber       *string  `json:"serialNumber,omitempty"`
	MACAddress         *string  `json:"macAddress,omitempty"`
	OperatingFrequency *float64 `json:"operatingFrequency,omitempty"`
	FrequencyRange     *float64 `json:"frequencyRange,omitempty"`
}
